HPCC-15180 Add eclcc option to save field usage

ericsuhong opened the pull request:
1. Added "recordFieldUsage" option to save field usage info into WU and Dali
2. Added an iterator for iterating through field usages:
   - can query file source type ("dataset" or "index")
   - can query file name
   - can query number of fields and used fields
   - can query field name

Signed-off-by: Suk Hwan Hong suk.hong@gatech.edu

ericsuhong committed:
HPCC-15180 Add eclcc option to save field usage

1. Added "recordFieldUsage" option to save field usage info into WU and Dali

2. Added an iterator for iterating through field usages:
    - can query file source type ("dataset" or "index")
    - can query file name
    - can query number of fields and used fields
    - can query field name

Signed-off-by: Suk Hwan Hong <suk.hong@gatech.edu>

sukhong commented:
@RussWhitehead Please Review

HPCCSmoketest commented:
Automated Smoketest 
Sha: a1a44f334bad66cb18151e1bdff1149720a95b38
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest 
Sha: 372c301d7503e0aa2b8a7e842a5758de98cd69ed
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

RussWhitehead commented on common/workunit/workunit.cpp:
```diff
@@ -1888,6 +1892,21 @@ class CConstWUArrayIterator : public CInterface, implements IConstWorkUnitIterat
+   bool                 first() { item.clear(); return iter->first(); }
+   bool                 isValid() { return iter->isValid(); }
+   bool                 next() { item.clear(); return iter->next(); }
+   IConstWUFieldUsageFile & query() { if (!item) item.setown(new CLocalWUFieldUsageFile(iter->get())); return *item; };
```
might be nice to add the override keyword to this method

RussWhitehead replied:
Would it make sense to create the CLocalWUFieldUsageFile instance in the CTOR instead of here?  Then the query method wont have to make the test with every invocation

sukhong replied:
Added override keyword.

For second question, I am following a pattern from CWorkflowItemIterator.

RussWhitehead replied:
I agree that the legacy code does it this way, but IMHO it's not right and slows every invocation of the method, both here and in the field iterator

sukhong replied:
Made a change to overall structure (removed cur, and get() method instead of query())

RussWhitehead commented on common/workunit/workunit.cpp:
```diff
@@ -1888,6 +1892,21 @@ class CConstWUArrayIterator : public CInterface, implements IConstWorkUnitIterat
+   IConstWUFieldUsageFile & query() { if (!item) item.setown(new CLocalWUFieldUsageFile(iter->get())); return *item; };
+private:
+   Owned<IPropertyTreeIterator> iter;
+   mutable Owned<CLocalWUFieldUsageFile> item;
```
Not sure why this is mutable

sukhong replied:
I missed const keyword for query() function. mutable is now needed with const query() function.

RussWhitehead commented on common/workunit/workunit.ipp:
```diff
@@ -87,6 +86,20 @@ public:
+    IMPLEMENT_IINTERFACE;
+    CLocalWUFieldUsageFile(IPropertyTree& _p) { p.setown(&_p); }
+
+    virtual IStringVal & getName(IStringVal & ret) const { ret.set(p->queryProp("@name")); return ret; }
```
another good set of places for "override"

sukhong replied:
No need to add "override" here since these member functions are new and do not override anything.

RussWhitehead commented on common/workunit/workunit.ipp:
```diff
@@ -87,6 +86,20 @@ public:
+    virtual IStringVal & getType(IStringVal & ret) const { ret.set(p->queryName()); return ret; }
+    virtual unsigned getNumFields() const { return p->getPropInt("@numFields"); }
+    virtual unsigned getNumFieldsUsed() const { return p->getPropInt("@numFieldsUsed"); }
+    virtual IPropertyTreeIterator * getFields() const { return p->getElements("field"); }
```
This is okay for now (IPropertyTreeIterator, but I have a feeling we will end up converting it to an IConstWUFieldUsageFieldIterator before we are done

sukhong replied:
Added an iterator for fields as well.

RussWhitehead commented on common/workunit/workunit.cpp:
```diff
@@ -6260,6 +6281,15 @@ void CLocalWorkUnit::noteFileRead(IDistributedFile *file)
+    if (fieldUsage)
+    {
+        CriticalBlock block(crit);
+        p->addPropTree("usedsources", LINK(fieldUsage));
```
Does fieldUsage need to be linked?  I think this might introduce a leak

sukhong replied:
fieldUsage solely is owned by enclosing generateStatistics function, and will be deallocated as soong as we exit the function.

However, field usage is not actually copied down to WU until later point, so we need to LINK the field usage to keep it in memory (within p) and avoid premature deallocation. 

field usage will be deallocated properly when its owning tree, p, is deallocated when its owning WU is destructed.

However, I moved the LINK() out of noteFieldUsage() and put it in enclosing generateStatistics() function to semantically indicate that noteFieldUsage() function will use passed fieldUsage at later point, and need not be deallocated immediately.

RussWhitehead replied:
Sounds good. At some point when this is functional we will run it with Valgrind to find any leaks we have introduced

RussWhitehead commented on ecl/hqlcpp/hqlcpp.cpp:
```diff
@@ -1750,6 +1750,7 @@ void HqlCppTranslator::cacheOptions()
         DebugOption(options.implicitGroupHashDedup,"implicitGroupHashDedup",false),
         DebugOption(options.reportFieldUsage,"reportFieldUsage",false),
         DebugOption(options.reportFileUsage,"reportFileUsage",false),
+        DebugOption(options.recordFieldUsage,"recordFieldUsage",false),
```
We need to remember to document this new #option as an ECL requirement for workunits that will be controlled by column level security

RussWhitehead commented on ecl/hqlcpp/hqlcpp.ipp:
```diff
@@ -733,6 +733,7 @@ struct HqlCppOptions
     bool                implicitGroupHashDedup;
     bool                reportFieldUsage;
     bool                reportFileUsage;
+    bool                recordFieldUsage; // enables column-level security when turned on
```
I would not mention column-level security in the comments throughout this PR. CLS is just one usage of recorded fieldUsage, there could be more. Maybe "enables recording field usage into the workunit"

sukhong replied:
Removed the comment

sukhong replied:
Removed the comment

RussWhitehead commented on common/workunit/workunit.cpp:
```diff
@@ -5231,6 +5250,8 @@ void CLocalWorkUnit::copyWorkUnit(IConstWorkUnit *cached, bool all)
         result.setResultStatus(ResultStatusUndefined);
     }
+
+    copyTree(p, fromP, "usedsources"); // field usage info for enforcing column-level security
```
Where in the workunit XML will this be?  I presume a branch off the root of the WU?

sukhong replied:
Correct.

RussWhitehead commented:
@sukhong Back at you

sukhong commented:
@RussWhitehead Please Review

HPCCSmoketest commented:
Automated Smoketest 
Sha: 5665a00e3cd52fdc4d1739c8531738a54aec2e24
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

RussWhitehead commented:
@sukhong Back at you

sukhong commented:
@RussWhitehead Please review

HPCCSmoketest commented:
Automated Smoketest 
Sha: ca925b593a82dc57caee723fbb13c37199a92d2d
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

RussWhitehead commented:
Looks good

sukhong commented:
@richardkchapman Please merge

richardkchapman commented on common/workunit/workunit.cpp:
```diff
@@ -5231,6 +5288,8 @@ void CLocalWorkUnit::copyWorkUnit(IConstWorkUnit *cached, bool all)
         result.setResultStatus(ResultStatusUndefined);
     }
+
+    copyTree(p, fromP, "usedsources"); // field usage
```
Not sure if this will work with Cassandra workunits (or at least, work would be needed in cassandrawu.cpp)

Are we sure that the workunit is the right place to store this information?

RussWhitehead replied:
@richardkchapman   ESP services will use this information to verify a user has access to all the fields before we submit the WU. I don't want to put it in the helper SO because then we have to wait until runtime, and the engines don's access LDAP anyway. I think we should make sure Cassandra supports this feature

RussWhitehead replied:
@sukhong Please look into \HPCC-Platform\plugins\cassandra\cassandrawu.cpp as Richard suggested

ericsuhong committed:
Added Field Usage support for Cassandra Workunits

Fixed xml format for fields usage, so that it is easier to map to cassandra tables

Added a suggestion to run wutool for an exception what occurs when a child table is missing

Signed-off-by: Suk Hwan Hong <suk.hong@gatech.edu>

HPCCSmoketest commented:
Automated Smoketest 
Sha: 8bbe9ab13f68a78c924bcd69c50dbceb06ba0014
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

sukhong commented:
@RussWhitehead Added Field Usage Support for Cassandra workunits. Please Review.

HPCCSmoketest commented:
Automated Smoketest 
Sha: f8be5fc1a7683040dd65e854609f93cf70d61091
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

RussWhitehead commented:
@sukhong Looks good, but lets have @richardkchapman review the cassandrawu.cpp changes as well

HPCCSmoketest commented:
Automated Smoketest 
Sha: f8be5fc1a7683040dd65e854609f93cf70d61091
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

sukhong commented:
@richardkchapman Please Merge

richardkchapman commented:
The Cassandra code looks ok, but I'm slightly unsure what will happen if you try to store field info on a system connecting to an older version of the store - the table is not going to be created except on Dali startup.

This may be something that requires documentation/RedBook rather than a code fix though

richardkchapman commented:
@sukhong One question before I pull...

HPCCSmoketest commented:
Automated Smoketest 
Sha: abdc03db82d3f4b5cf7caf8694830c8ae727ec8f
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

sukhong commented:
@richardkchapman 
Currently, if you try to store fields info on a system connecting to an older version of the store, it will throw "cassandra: failed to executeQuery (All hosts in current policy attempted and were either unavailable or failed)" without any specific reason.

I've found two places (in checkChildLoaded() and in childXMLRowtoCassandra()) where exception may occur when a child table is missing. In those places, I wrapped an exception to throw a more meaningful error **with a suggestion to restart Dali**:

> "Failed to load cassandra child table 'wuFieldUsage' (Dali restart may be required to setup all cassandra tables), [cassandra: failed to executeQuery (All hosts in current policy attempted and were either unavailable or failed)]."

I've appended the original error message at the end, just in case.

I think these measures will be sufficient for minimizing surprises (and give self-guidance) to admins in the future.

We will definitely add documentations regarding field usage, and there is a Jira ticket assigned to it (HPCC-15752). We will clearly mention this case of upgrading the system.

sukhong commented:
@richardkchapman Please Review

HPCCSmoketest commented:
Automated Smoketest 
Sha: 344791bc5a835b02f5960ba02611c320af51cc38
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest 
Sha: d78bebe73401afaefcaea378ba5ce7403f83e63c
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest 
Sha: 2d802f361dc5cfb8090e9e506100306fe3c7053f
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest 
Sha: bab147975dbadbf5200f84a81be8aed0ce9c1f03
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

richardkchapman commented:
It's not really a restart of dali that is needed, so much as an upgrade of dali to the latest version. Which history suggests ops will not always be willing/able to do at short notice.

It's probably also possible to run wutool to create the required tables (and if not, it should be!)

sukhong commented:
@richardkchapman One question I have is, is it a common scenario that an admin upgrades/restarts the whole system except Dali? If Dali is not restarted (but may have been upgraded), but other systems are upgraded and restarted, isn't it possible that other compatibility issues will arise as well?

I will start looking at the wutool!

HPCCSmoketest commented:
Automated Smoketest 
Sha: 57f7293abb41c1533e5685ca3b3e644f41fa10fe
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

richardkchapman commented:
@ghalliday Have you looked at this?

ghalliday commented on ecl/hql/hqlusage.cpp:
```diff
@@ -416,8 +417,9 @@ IPropertyTree * SourceFieldUsage::createReport(bool includeFieldDetail, const IP
     }
 
-    Owned<IPropertyTree> entry = createPTree(type);
+    Owned<IPropertyTree> entry = createPTree("datasource");
```
This format change will need documenting - anyone who is currently reading the associated field information will need to know that this change is coming.

sukhong replied:
This is changed so that it can easily be mapped to cassandra tables using existing cassandra "mapping" code.

ghalliday commented on ecl/hql/hqlusage.cpp:
```diff
@@ -344,6 +344,7 @@ static IPropertyTree * addSelect(IPropertyTree * xml, IHqlExpression * expr, boo
     Owned<IPropertyTree> field = createPTree(isUsed ? "field" : "unused");
     field->setProp("@name", text.str());
     const char * tag = field->queryName();
+    xml = ensurePTree(xml, "fields");
```
This is another format change.  Is it required?  Is it due to the cassandra integration?

sukhong replied:
This is changed so that it can be easily mapped to cassandra tables using existing cassandra "mapping" code.

ghalliday commented:
@sukhong This looks good to merge, but I have a couple of questions about the changes to the format.

sukhong commented:
@ghalliday Two format changes are due to cassandra integration. I tried hard not to modify existing XML format, but I deemed it necessary for easy mapping with cassandra without introducing any complicated code in cassandrawu.cpp. 

In order to map fieldusage to cassandra tables easily using existing cassandra "mapping" code, following changes were made:

1. All used datasources are created with a common <datasource> tag, and their types are recorded as a attribute "type" (dataset or index):
<dataset name="test" numFields=10 numFieldsUsed=10></dataset>
    =>
<datasource name="test" numFields=10 numFieldsUsed=10 type="dataset"></dataset>

2. All used fields are wrapped in additional <fields> tag:
<datasource><field name="firstname"/><field name="lastname"/></datasource>
    =>
<datasource><fields><field name="firstname"/><field name="lastname"/></fields></datasource>

We will certainly document these changes along with other new changes ("recordFieldUsage" option and etc).

ericsuhong committed:
FieldUsage Logic Changes

- Empty fieldusage is added for a query that is not using any fields
- reportFieldUsage and recordFieldUsage can happen at the same time
- added a helper method to get string array for field usage
- default scope is loaded and prefixed to a filename if filename
  doesn't start with a leading tilde

Signed-off-by: Suk Hwan Hong <suk.hong@gatech.edu>

sukhong commented:
@ghalliday Also made one more change that adds empty field usage for a query that does not use any fields. 

This change is made to differentiate between a query that was not compiled with a "recordFieldUsage" option (will not have <usedsources> tag) and a query that does not use any fields (will have empty <usedsources> tag).

HPCCSmoketest commented:
Automated Smoketest 
Sha: ceb0e9d7f7d5a4b19c492f181006991b3da9d2ca
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest 
Sha: da9990fa94d0c12bad0a259f39d4e6f1bbd899ae
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK

sukhong commented:
@ghalliday Please merge

HPCCSmoketest commented:
Automated Smoketest 
Sha: 2062654481b122495b54f63ee2d449957b520ee6
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
Unittest result: total:82 passed:82 failed:0

HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest 
Sha: 2062654481b122495b54f63ee2d449957b520ee6
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
Unittest result: total:82 passed:82 failed:0

HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest 
Sha: c0d13c39676da81744184e9b5b0b4d49c8876172
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
Unittest result: total:82 passed:82 failed:0

HPCC Stop: OK
HPCC Uninstall: OK

sukhong commented:
@ghalliday Please review/merge. Squashed all changes since the last review into one commit.

ghalliday commented on common/workunit/workunit.cpp:
```diff
@@ -6359,6 +6429,83 @@ unsigned CLocalWorkUnit::queryFileUsage(const char *fileName) const
+        char* cleanFilename = new char[filename.length()-1];
+        memcpy(cleanFilename, filename.str()+1, filename.length()-2);
+        cleanFilename[filename.length()-2] = '\0';
+        filename.set(cleanFilename); 
```
easier to say

StringAttr cleanFilename(cleanFilename.str()+1, cleanFilename.length()-2);

sukhong replied:
Fixed.

ghalliday commented on common/workunit/workunit.cpp:
```diff
@@ -6359,6 +6429,83 @@ unsigned CLocalWorkUnit::queryFileUsage(const char *fileName) const
+        if (filename.length() < 3)
+            throw MakeStringException(WUERR_InvalidFieldUsage, "Invalid FieldUsage found in WU. Cannot enforce view security.");
+
+        // Filenames in field usage are surrounded with single quotes ('filename').
```
Why are they surrounded with single quotes?  That sounds strange.

ghalliday replied:
Actually I understand why - because it is the ecl expression.

However the filenames may also be variable - in which case they are not known at compile time.  What should the code do in that situation?  It should verify that it is surrounded by quotes.  If not, should this fail?

sukhong replied:
I made a fix so that when a filename is not surrounded by a single quote, it is not treated as a resolved filename. When a query contains any unresolved filename, it means a query may potentially access ANY files, so it will be denied unconditionally (in terms of view security).

ghalliday commented on common/workunit/workunit.hpp:
```diff
@@ -378,6 +378,29 @@ interface IConstWUAssociatedFileIterator : extends IScmIterator
 
+interface IConstWUFieldUsage : extends IInterface // Defines a file (dataset or index) that contains used fields from queries
+{
+    virtual IStringVal & getName(IStringVal & ret) const = 0;
```
minor:  For future reference, there has been a move away from functions like this to functions:

    virtual const char * queryName() const = 0;

because it can be implemented more efficiently.  

E.g.,     virtual const char *queryWuid() const = 0; in IConstWorkUnitInfo

sukhong replied:
Fixed.

ghalliday commented:
@sukhong a couple of final comments to do with quoted filenames.

ericsuhong committed:
code review fix

Signed-off-by: Suk Hwan Hong <suk.hong@gatech.edu>

sukhong commented:
@ghalliday Please review again.

HPCCSmoketest commented:
Automated Smoketest 
Sha: 4e124afc12ba28e1ec924740f74b3e6e43b47917
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
Unittest result: total:82 passed:82 failed:0

HPCC Stop: OK
HPCC Uninstall: OK
